### PR TITLE
[16.0][ADD] stock_release_channel_delivery

### PR DESCRIPTION
**Pull request by trisdoan**
This addon adds a selection criteria on the release channel based on the shipping
method.

### CODE REVIEW

**Comment by jbaudoux** [comment]
@trisdoan Can you rewrite commit msg
`[IMP] implement JE's change requests`
to 
```
[IMP] stock_release_channel_delivery:

allow to configure multiple shipping methods on a channel
```

**Comment by jbaudoux** [comment]
Note: failing test is not related to this PR

File: stock_release_channel_delivery/i18n/fr.po [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+#. module: stock_release_channel_delivery
+#: model:ir.model.fields,help:stock_release_channel_delivery.field_stock_release_channel__carrier_id
+msgid "The carrier linked to the release channel."
+msgstr "Le transporteur associé au release channel"
+
+#. module: stock_release_channel_delivery
+#: model:ir.model.fields,field_description:stock_release_channel_delivery.field_stock_release_channel__carrier_id
+msgid "Transporter"
+msgstr "Transporteur"
```
Comment (jbaudoux) [inline stock_release_channel_delivery/i18n/fr.po]:
```suggestion
#: model:ir.model.fields,field_description:stock_release_channel_delivery.field_stock_release_channel__carrier_ids
msgid "Shipping Methods"
msgstr "Méthodes de livraison"
```

File: stock_release_channel_delivery/i18n/fr.po [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+#. module: stock_release_channel_delivery
+#: model:ir.model,name:stock_release_channel_delivery.model_stock_release_channel
+msgid "Stock Release Channels"
+msgstr ""
+
+#. module: stock_release_channel_delivery
+#: model:ir.model.fields,help:stock_release_channel_delivery.field_stock_release_channel__carrier_id
+msgid "The carrier linked to the release channel."
+msgstr "Le transporteur associé au release channel"
```
Comment (jbaudoux) [inline stock_release_channel_delivery/i18n/fr.po]:
```suggestion
#: model:ir.model.fields,help:stock_release_channel_delivery.field_stock_release_channel__carrier_ids
msgid "The shipping methods linked to the release channel."
msgstr "Les méthodes de livraisons associées au release channel"
```

File: stock_release_channel_delivery/i18n/stock_release_channel_delivery.pot [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+"Plural-Forms: \n"
+
+#. module: stock_release_channel_delivery
+#: model:ir.model,name:stock_release_channel_delivery.model_stock_release_channel
+msgid "Stock Release Channels"
+msgstr ""
+
+#. module: stock_release_channel_delivery
+#: model:ir.model.fields,help:stock_release_channel_delivery.field_stock_release_channel__carrier_id
+msgid "The carrier linked to the release channel."
```
Comment (jbaudoux) [inline stock_release_channel_delivery/i18n/stock_release_channel_delivery.pot]:
```suggestion
#: model:ir.model.fields,help:stock_release_channel_delivery.field_stock_release_channel__carrier_ids
msgid "The shipping methods linked to the release channel."
```

File: stock_release_channel_delivery/i18n/stock_release_channel_delivery.pot [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+msgstr ""
+
+#. module: stock_release_channel_delivery
+#: model:ir.model.fields,field_description:stock_release_channel_delivery.field_stock_release_channel__carrier_id
+msgid "Transporter"
```
Comment (jbaudoux) [inline stock_release_channel_delivery/i18n/stock_release_channel_delivery.pot]:
```suggestion
#: model:ir.model.fields,field_description:stock_release_channel_delivery.field_stock_release_channel__carrier_ids
msgid "Shipping Methods"
```

**Review by jbaudoux — COMMENTED** [review]
LG.
Can you also fix the translations

**Comment by sebalix** [comment]
Build fixed here: https://github.com/OCA/wms/pull/849

**Comment by jbaudoux** [comment]
@trisdoan Can you rebase this one ?

**Comment by trisdoan** [comment]
Hello @jbaudoux, it's done

**Review by jbaudoux — APPROVED** [review]

**Comment by jbaudoux** [comment]
Thanks

**Comment by jbaudoux** [comment]
ping @rousseldenis @lmignon

File: stock_release_channel_delivery/migrations/16.0.2.0.0/pre-migration.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# Copyright 2024 Camptocamp
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openupgradelib import openupgrade
+
+
+@openupgrade.migrate()
+def migrate(env, version):
+    if openupgrade.column_exists(env.cr, "stock_release_channel", "carrier_id"):
```
Comment (rousseldenis) [inline stock_release_channel_delivery/migrations/16.0.2.0.0/pre-migration.py:9]: @trisdoan @jbaudoux Doesn't this conflict with post script ?

File: stock_release_channel_delivery/models/stock_picking.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    _inherit = "stock.picking"
+
+    def _get_release_channel_possible_candidate_domain(self):
+        domain = super()._get_release_channel_possible_candidate_domain()
+        if self.carrier_id:
+            domain.extend(
+                [
+                    "|",
+                    ("carrier_ids", "=", False),
+                    ("carrier_ids", "=", self.carrier_id.ids),
```
Comment (rousseldenis) [inline stock_release_channel_delivery/models/stock_picking.py]: Don't you have a warning as "in" is better IMHO?

Comment (rousseldenis) [reply stock_release_channel_delivery/models/stock_picking.py]: Indeed, you have:

image

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review

File: stock_release_channel_delivery/migrations/16.0.2.0.0/pre-migration.py (hunk 5 above)
Comment (jbaudoux) [reply stock_release_channel_delivery/migrations/16.0.2.0.0/pre-migration.py:9]: That's per documentation in https://oca.github.io/openupgradelib/API.html#openupgradelib.openupgrade.m2o_to_x2m

How should it be according to you?

Comment (rousseldenis) [reply stock_release_channel_delivery/migrations/16.0.2.0.0/pre-migration.py:9]: But you rename the column to nothing ?

Comment (rousseldenis) [reply stock_release_channel_delivery/migrations/16.0.2.0.0/pre-migration.py:9]: Ooooh, ok didn't know the legacy name

**Comment by trisdoan** [comment]
Hello @rousseldenis, thank you for your in-depth review.

I updated as you suggested.

**Review by cyrilmanuel — APPROVED** [review]
Not an expert but LGTM

@rousseldenis  little ping, is it good for you ?

**Review by lmignon — APPROVED** [review]
LGTM (Code review)

**Pull request merged**